File: ai-agent/bot.js
import TelegramBot from 'node-telegram-bot-api';
import dotenv from 'dotenv';
import { getLatestData, getDataByDate, getSummary } from './db.js';
import setupNLP, { formatResponse } from './nlp-trainer.js';

dotenv.config();

const token = process.env.TELEGRAM_TOKEN;
let nlp;

// Inisialisasi bot
const bot = new TelegramBot(token, { polling: true });

// Helper function untuk escape karakter Markdown
function escapeMarkdown(text) {
  return text.toString().replace(/[_*[\]()~`>#+=|{}.!-]/g, '\\$&');
}

// Parse tanggal dari teks
function parseDateFromText(text) {
  const dateRegex = /(\d{1,2})[\/\-](\d{1,2})[\/\-]?(\d{4})?/;
  const match = text.match(dateRegex);
  
  if (match) {
    const day = parseInt(match[1]);
    const month = parseInt(match[2]);
    const year = match[3] ? parseInt(match[3]) : new Date().getFullYear();
    return new Date(year, month - 1, day);
  }
  
  return null;
}

// Format respons sensor untuk Telegram
function formatSensorResponse(data) {
  if (!data || !data.length) return '❌ Tidak ada data tersedia';
  
  let response = '📈 *Data Sensor Terkini*\n\n';
  data.slice(0, 3).forEach(row => {
    response += `⏰ *${escapeMarkdown(row.timestamp)}*\n`;
    response += `• Suhu: *${escapeMarkdown(row.suhu)}°C*\n`;
    response += `• Kelembaban: *${escapeMarkdown(row.kelembaban)}%*\n`;
    response += `• pH: *${escapeMarkdown(row.ph)}*\n`;
    response += `• Nutrisi: *${escapeMarkdown(row.nutrisi)} PPM*\n`;
    response += `• Status: _${escapeMarkdown(row.status_tanaman)}_\n\n`;
  });
  
  return response;
}

// Format ringkasan untuk Telegram
function formatSummaryResponse(data) {
  if (!data) return '❌ Tidak ada data tersedia';
  
  return `📊 *Ringkasan Kondisi Tanaman*

• Suhu: *${escapeMarkdown(data.suhuRataRata)}°C*
• Kelembaban: *${escapeMarkdown(data.kelembabanRataRata)}%*
• pH: *${escapeMarkdown(data.phRataRata)}*
• Nutrisi: *${escapeMarkdown(data.nutrisiRataRata)} PPM*

_Data dari ${escapeMarkdown(data.totalData)} pengukuran terakhir_`;
}

// Handle pesan masuk
async function handleMessage(msg) {
  const chatId = msg.chat.id;
  const text = msg.text.toLowerCase();

  try {
    // Analisis teks menggunakan NLP
    const result = await nlp.process('id', text);
    const intent = result.intent;

    // Handle berdasarkan intent
    switch (intent) {
      case 'greeting':
        const welcomeText = `🌱 *Selamat datang di Hidroponik AI Assistant\\!*

Saya dapat membantu Anda memantau sistem hidroponik\\. Anda bisa bertanya tentang:
• Status dan kondisi tanaman
• Data sensor terbaru
• Analisis kelembaban dan nutrisi
• Data berdasarkan tanggal
• Dan informasi lainnya

_Silakan ajukan pertanyaan Anda_ 😊`;
        await bot.sendMessage(chatId, welcomeText, { parse_mode: 'MarkdownV2' });
        break;

      case 'check_status':
        const summary = await getSummary();
        await bot.sendMessage(chatId, formatSummaryResponse(summary), { parse_mode: 'MarkdownV2' });
        break;

      case 'show_data':
        const data = await getLatestData(3);
        await bot.sendMessage(chatId, formatSensorResponse(data), { parse_mode: 'MarkdownV2' });
        break;

      case 'check_temperature':
        const tempData = await getLatestData(1);
        if (tempData.length > 0) {
          await bot.sendMessage(
            chatId,
            `🌡️ *Suhu saat ini:* ${escapeMarkdown(tempData[0].suhu)}°C`,
            { parse_mode: 'MarkdownV2' }
          );
        }
        break;

      case 'check_humidity':
        const humidData = await getLatestData(1);
        if (humidData.length > 0) {
          await bot.sendMessage(
            chatId,
            `💧 *Kelembaban saat ini:* ${escapeMarkdown(humidData[0].kelembaban)}%`,
            { parse_mode: 'MarkdownV2' }
          );
        }
        break;

      case 'check_ph':
        const phData = await getLatestData(1);
        if (phData.length > 0) {
          await bot.sendMessage(
            chatId,
            `🧪 *pH saat ini:* ${escapeMarkdown(phData[0].ph)}`,
            { parse_mode: 'MarkdownV2' }
          );
        }
        break;

      case 'check_nutrients':
        const nutrientData = await getLatestData(1);
        if (nutrientData.length > 0) {
          await bot.sendMessage(
            chatId,
            `🌿 *Nutrisi saat ini:* ${escapeMarkdown(nutrientData[0].nutrisi)} PPM`,
            { parse_mode: 'MarkdownV2' }
          );
        }
        break;

      case 'check_date':
        const date = parseDateFromText(text);
        if (date) {
          const dateData = await getDataByDate(date);
          await bot.sendMessage(chatId, formatSensorResponse(dateData), { parse_mode: 'MarkdownV2' });
        } else {
          await bot.sendMessage(
            chatId,
            '❌ Format tanggal tidak valid\\. Gunakan format DD/MM/YYYY',
            { parse_mode: 'MarkdownV2' }
          );
        }
        break;

      default:
        await bot.sendMessage(
          chatId,
          'Maaf, saya tidak mengerti pertanyaan tersebut\\. Silakan coba dengan kata kunci lain\\.',
          { parse_mode: 'MarkdownV2' }
        );
    }
  } catch (error) {
    console.error('Error:', error);
    await bot.sendMessage(
      chatId,
      '❌ Terjadi kesalahan saat memproses pesan\\. Silakan coba lagi\\.',
      { parse_mode: 'MarkdownV2' }
    );
  }
}

// Setup bot dan NLP
async function startBot() {
  try {
    // Setup NLP
    nlp = await setupNLP();
    console.log('NLP model trained and ready');

    // Listen for messages
    bot.on('message', handleMessage);
    console.log('Bot is running...');
  } catch (error) {
    console.error('Error starting bot:', error);
  }
}

export default startBot;

File: ai-agent/bot.py
import os
from telegram import Update
from telegram.constants import ParseMode
from telegram.ext import Application, MessageHandler, filters, ContextTypes
from dotenv import load_dotenv
from main import analyze_hydroponic_data
from db_loader import HydroponicDBLoader
from telegram_helper import (
    escape_markdown,
    format_sensor_data,
    format_summary,
    parse_date_query,
    safe_markdown_message
)

# Load environment variables
load_dotenv()
TELEGRAM_TOKEN = os.getenv("TELEGRAM_TOKEN")

# Initialize database loader
db_loader = HydroponicDBLoader()

def get_data_summary(date=None):
    """Get formatted summary of recent data."""
    try:
        summary = db_loader.get_latest_summary()
        docs = db_loader.load_data()
        
        if date:
            # Filter documents by date if specified
            docs = [doc for doc in docs if date.date() == doc.metadata.get('timestamp').date()]
            if not docs:
                return f"❌ Tidak ada data untuk tanggal {date.strftime('%d %B %Y')}"
        
        response = f"""
📊 *Ringkasan Data Hidroponik*

{format_summary(summary)}

📝 *Data Terakhir:* _{len(docs)} record_
"""
        # Format each sensor reading
        for i, doc in enumerate(docs[:3]):
            try:
                # Parse document content into dictionary
                data = {
                    'timestamp': doc.metadata.get('timestamp'),
                    'suhu': doc.metadata.get('suhu'),
                    'kelembaban': doc.metadata.get('kelembaban'),
                    'ph': doc.metadata.get('ph'),
                    'nutrisi': doc.metadata.get('nutrisi'),
                    'status': doc.metadata.get('status', 'Normal')
                }
                response += format_sensor_data(data)
            except Exception:
                response += f"\n{escape_markdown(doc.page_content)}"
        
        if len(docs) > 3:
            response += f"\n\n_{escape_markdown(f'...dan {len(docs)-3} data lainnya')}_"
            
        return response
    except Exception as e:
        return f"❌ Error mengambil data: {escape_markdown(str(e))}"

async def handle_message(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Handle all messages with natural language processing."""
    try:
        message = update.message.text.lower()
        
        # Pesan selamat datang
        if any(word in message for word in ['halo', 'hai', 'hi', 'hello']):
            welcome_text = """
🌱 *Selamat datang di Hidroponik AI Assistant\!*

Saya dapat membantu Anda memantau sistem hidroponik\. Anda bisa bertanya tentang:
• Status dan kondisi tanaman
• Data sensor terbaru
• Analisis kelembaban dan nutrisi
• Rekomendasi perawatan
• Data berdasarkan tanggal tertentu
• Dan informasi lainnya

_Silakan ajukan pertanyaan Anda_ 😊
"""
            await update.message.reply_text(welcome_text, parse_mode=ParseMode.MARKDOWN)
            return

        # Cek query tanggal
        date = parse_date_query(message)
        if date:
            response = get_data_summary(date)
            await update.message.reply_text(response, parse_mode=ParseMode.MARKDOWN)
            return

        # Cek jika user menanyakan tentang data
        if any(phrase in message for phrase in ['data', 'sensor', 'berapa data', 'lihat data']):
            response = get_data_summary()
            await update.message.reply_text(response, parse_mode=ParseMode.MARKDOWN)
            return

        # Untuk pertanyaan lainnya, gunakan LangChain untuk analisis
        result = analyze_hydroponic_data(message)
        # Pastikan response aman untuk Markdown
        safe_result = safe_markdown_message(result)
        await update.message.reply_text(safe_result, parse_mode=ParseMode.MARKDOWN)

    except Exception as e:
        error_message = f"""
❌ *Maaf, terjadi error*

_{escape_markdown(str(e))}_

Silakan coba lagi atau ajukan pertanyaan dengan cara berbeda\.
"""
        # Jika masih error, kirim tanpa Markdown
        try:
            await update.message.reply_text(error_message, parse_mode=ParseMode.MARKDOWN)
        except Exception:
            await update.message.reply_text(f"Error: {str(e)}")

def main():
    """Start the bot."""
    # Create the Application
    application = Application.builder().token(TELEGRAM_TOKEN).build()

    # Handle all messages
    application.add_handler(MessageHandler(filters.TEXT, handle_message))

    # Start the bot
    print("Bot started... Ready to process natural language queries!")
    application.run_polling(allowed_updates=Update.ALL_TYPES)

if __name__ == '__main__':
    main()
